security/advancedtls/crl.go

```diff
@@ -0,0 +1,499 @@
+			} else if len(rest) != 0 {
+				return nil, errors.New("trailing data after IssuingDistributionPoint extension")
+			}
+
+			if dp.OnlyContainsUserCerts || dp.OnlyContainsCACerts || dp.OnlyContainsAttributeCerts {
+				return nil, errors.New("CRL only contains some certificate types")
+			}
+			if dp.IndirectCRL {
+				return nil, errors.New("indirect CRLs unsupported")
+			}
+			if dp.OnlySomeReasons.BitLength != 0 {
+				return nil, errors.New("onlySomeReasons unsupported")
+			}
+
+		case ext.Critical:
+			return nil, fmt.Errorf("unsupported critical extension: %v", ext.Id)
+		}
+	}
+
+	if len(certList.AuthorityKeyID) == 0 {
+		return nil, errors.New("authority key identifier extension missing")
+	}
+	return certList, nil
+}
+
+func fetchCRL(loc string, rawIssuer []byte, cfg RevocationConfig) (*certificateListExt, error) {
+	var parsedCRL *certificateListExt
+	// 6.3.3 (a) (1) (ii)
+	// According to X509_LOOKUP_hash_dir the format is issuer_hash.rN where N is an increasing number.
+	// There are no gaps, so we break when we can't find a file.
+	for i := 0; ; i++ {
+		// Unmarshal to RDNSeqence according to http://go/godoc/crypto/x509/pkix/#Name.
+		var r pkix.RDNSequence
+		rest, err := asn1.Unmarshal(rawIssuer, &r)
+		if len(rest) != 0 || err != nil {
+			return nil, fmt.Errorf("asn1.Unmarshal(Issuer) len(rest) = %v, err = %v", len(rest), err)
+		}
+		crlPath := fmt.Sprintf("%s.r%d", filepath.Join(cfg.RootDir, x509NameHash(r)), i)
+		crlBytes, err := ioutil.ReadFile(crlPath)
+		if err != nil {
+			// Break when we can't read a CRL file.
+			grpclogLogger.Infof("readFile: %v", err)
+			break
+		}
+
+		crl, err := x509.ParseCRL(crlBytes)
+		if err != nil {
+			// Parsing errors for a CRL shouldn't happen so fail.
+			return nil, fmt.Errorf("x509.ParseCrl(%v) failed err = %v", crlPath, err)
+		}
+		var certList *certificateListExt
+		if certList, err = parseCRLExtensions(crl); err != nil {
+			grpclogLogger.Infof("fetchCRL: unsupported crl %v, err = %v", crlPath, err)
+			// Continue to find a supported CRL
+			continue
+		}
+
+		rawCRLIssuer, err := asn1.Marshal(certList.CertList.TBSCertList.Issuer)
+		if err != nil {
+			return nil, err
+		}
+		// RFC5280, 6.3.3 (b) Verify the issuer and scope of the complete CRL.
+		if bytes.Equal(rawIssuer, rawCRLIssuer) {
+			parsedCRL = certList
+			// Continue to find the highest number in the .rN suffix.
+			continue
+		}
+	}
+
+	if parsedCRL == nil {
+		return nil, fmt.Errorf("fetchCrls no CRLs found for issuer")
+	}
+	return parsedCRL, nil
+}
+
+func verifyCRL(crl *certificateListExt, rawIssuer []byte, chain []*x509.Certificate) error {
+	// RFC5280, 6.3.3 (f) Obtain and validateate the certification path for the issuer of the complete CRL
+	// We intentionally limit our CRLs to be signed with the same certificate path as the certificate
+	// so we can use the chain from the connection.
+	rawCRLIssuer, err := asn1.Marshal(crl.CertList.TBSCertList.Issuer)
+	if err != nil {
+		return err
```

Comment:
This should probably include more info like "failed to marshal issuer <other info?>: ..."

This actually applies anywhere you do `return <nil/etc>, err`.  Just propagating an error verbatim is usually a bad idea since the context is lost.